### PR TITLE
[impeller] Implement an OpenGL ES 2.0 backend.

### PR DESCRIPTION
**Pull request by chinmaygarde**
This is the first pass at implementing an Impeller backend that restricts itself
to the feature set available in OpenGL ES 2.0. This means there is no use of
features that we will end up using where available. These features include
vertex array objects, uniform buffer objects, frame buffer discard operations,
etc..

As written, this backend should support 100% of the Android devices out in the
wild. Using VAO's and UBO's (in a future patch) where available should benefit
90.91% of the devices in the wild. Beyond that, I don't believe it makes too
much sense to invest in this backend.
Screen Shot 2022-05-03 at 2 32 38 PM

This was a fairly straightforward implementation of an Impeller backend with the
following items that are perhaps unique:
* Since there are no uniform buffer objects in ES2, the offline reflector in
  `impellerc` generates additional metadata for backends like ES2 to find and
  bind uniform data in existing device buffers. Commands reference this data in
  their stage bindings but only via a pointer to a const object. Backends that
  don't need this metadata don't see any additional overhead. In the OpenGL ES
  backend, deletion of the buffers that contain uniform data is delayed till
  uniform data is bound via the glUniform* family of functions.
* Unlike Skia objects that reference GPU resources, Impeller objects can be
  created, used, and collected on any thread. All OpenGL operations happen in
  the `impeller::ReactorGLES`. This makes it easy to deal with context loss as
  we don't have to chase references to resources or ensure resources are
  collected on the right thread. Impeller resources (like textures, pipelines,
  buffers, etc.) that reference GPU resources will just reference dead handles
  on context loss and can be collected when convenient. The reactor is still
  WIP and needs additional synchronization and tests for thread safety.
* Instead of not being able to use settings in high level API based on the
  backend in use, the specific settings (which were as yet unused) have been
  removed. This include `BlendOperation::kMin` and `BlendOperation::kMax` which
  require extensions in OpenGL ES 2.0. The addition of this new backend doesn't
  affect how the higher levels in the stack (like the renderer, entity, aiks,
  display_list, etc.) function.
* This backend performs an additional copy of buffer data since it need to push
  operations into the reactor. This copy will be removed when the appropriate
  APIs are updated ([in a later patch](https://github.com/flutter/flutter/issues/102262)) to use mappings and buffer views. Memory
  usage of this backend will go down significantly after that patch. Even more still
  on devices where UBO's are available.
* This backend will generate separate program objects for pipeline variants.
  This makes no sense for OpenGLES as that state is set separately per draw
  call. Reusing of the same program handle for pipeline variants is as yet
  unimplemented.
* The playground test harness that was previously parameterized now has the
  `PlaygroundBackend::kOpenGLES` parameter. This effectively doubles the
  playground tests corpus.
* There is error validation after each call to OpenGL. Ideally, this should be
  moved to debug runtime modes.
* Capturing RenderDoc traces is hard in the playground because both SwiftShader
  is also linked in. Devising a strategy for capturing RenderDoc traces from the
  playgrounds is pending.

TL;DR: Backend is ready for review. Needs debugging and additional refactors
(already listed) to bring memory usage back down to baseline (relative to the
Metal backend). Existing API and backends unaffected.
Screen Shot 2022-05-04 at 12 43 33 PM

### CODE REVIEW

File: impeller/playground/backend/gles/playground_impl_gles.cc [hunk 1]
```diff
@@ -3,35 +3,111 @@
...
+  ::glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 2);
+  ::glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 0);
+#endif  // FML_OS_MACOSX
+  ::glfwWindowHint(GLFW_RED_BITS, 8);
+  ::glfwWindowHint(GLFW_GREEN_BITS, 8);
+  ::glfwWindowHint(GLFW_BLUE_BITS, 8);
+  ::glfwWindowHint(GLFW_ALPHA_BITS, 8);
+  ::glfwWindowHint(GLFW_DEPTH_BITS, 0);    // no depth buffer
+  ::glfwWindowHint(GLFW_STENCIL_BITS, 8);  // 8 bit stencil buffer
+  ::glfwWindowHint(GLFW_SAMPLES, 0);       // 4xMSAA (disabled)
```
Comment (dnfield) [inline impeller/playground/backend/gles/playground_impl_gles.cc:44]: Why disable MSAA in the playground?

**Review by dnfield — APPROVED** [review]
This LGTM to land and iterate on. I have not closely read every line of the patch though.

**Comment by dnfield** [comment]
License bot is angry

File: impeller/playground/backend/gles/playground_impl_gles.cc (hunk 1 above)
Comment (chinmaygarde) [reply impeller/playground/backend/gles/playground_impl_gles.cc:44]: I haven't verified these work in the playground yet and I was trying to simplify the test case to make the frame debugger connect. I'll back this out.

File: impeller/renderer/backend/gles/formats_gles.h [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+  switch (type) {
+    case ShaderType::kSignedByte:
+      return GL_BYTE;
+    case ShaderType::kUnsignedByte:
+      return GL_UNSIGNED_BYTE;
+    case ShaderType::kSignedShort:
+      return GL_SHORT;
+    case ShaderType::kUnsignedShort:
+      return GL_UNSIGNED_SHORT;
+    case ShaderType::kUnknown:
```
Comment (chinmaygarde) [inline impeller/renderer/backend/gles/formats_gles.h:145]: This is incorrect.

File: impeller/renderer/backend/gles/vertex_descriptor_gles.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+#include "flutter/fml/macros.h"
+#include "impeller/renderer/backend/gles/gles.h"
+#include "impeller/renderer/backend/gles/proc_table_gles.h"
+#include "impeller/renderer/command.h"
+#include "impeller/renderer/vertex_descriptor.h"
+
+namespace impeller {
+
+class VertexDescriptorGLES {
```
Comment (chinmaygarde) [inline impeller/renderer/backend/gles/vertex_descriptor_gles.h:18]: This was modeled after VertexDescriptorMTL but is way more than just that. Rename class and dcoument.

File: impeller/renderer/command.h [hunk 4]
```diff
@@ -65,13 +81,53 @@ struct Command {
...
+  //----------------------------------------------------------------------------
+  /// The orientation of vertices of the front-facing polygons. This usually
+  /// matters when culling is enabled.
+  ///
+  /// @see         `cull_mode`
+  ///
   WindingOrder winding = WindingOrder::kClockwise;
+  //----------------------------------------------------------------------------
+  /// How to control culling of polygons. The orientation of front-facing
+  /// polygons is contrlled via the `winding` parameter.
```
Comment (chinmaygarde) [inline impeller/renderer/command.h:119]: Typo.

File: impeller/renderer/formats.h [hunk 5]
```diff
@@ -153,16 +174,21 @@ enum class PrimitiveType {
...
-  Scalar znear = 0.0f;
-  Scalar zfar = 1.0f;
+  DepthRange depth_range;
 };
 
 enum class MinMagFilter {
   /// Select nearest to the sample point. Most widely supported.
   kNearest,
-  /// Select two points and linearly interpolate between them. Some formats may
+  /// Select two points and linearly interpolate between them. Some formats
```
Comment (chinmaygarde) [inline impeller/renderer/formats.h:190]: Formatting error.

File: impeller/renderer/platform.h [hunk 6]
```diff
@@ -17,7 +17,7 @@ constexpr size_t DefaultUniformAlignment() {
 #elif FML_OS_MACOSX
   return 256u;
 #else
-#error "Unsupported platform".
+  return 16u;
```
Comment (chinmaygarde) [inline impeller/renderer/platform.h:20]: Make this backend dependent and tighten for client provided buffer data.

File: impeller/renderer/formats.h [hunk 7]
```diff
@@ -11,6 +11,7 @@
 
 #include "flutter/fml/hash_combine.h"
 #include "flutter/fml/macros.h"
+#include "fml/logging.h"
```
Comment (bdero) [inline impeller/renderer/formats.h:14]: The world's smallest nit: "flutter/fml/logging.h"

File: impeller/renderer/formats.h [hunk 8]
```diff
@@ -86,8 +87,6 @@ enum class BlendOperation {
   kAdd,
   kSubtract,
   kReverseSubtract,
-  kMin,
-  kMax,
```
Comment (bdero) [inline impeller/renderer/formats.h:90]: This reminds me: At some point we gotta talk about a strategy for extensions. We can make all the advanced blends done in-pipeline with [EXT_blend_minmax](https://www.khronos.org/registry/OpenGL/extensions/EXT/EXT_blend_minmax.txt) + [KHR_blend_equation_advanced](https://www.khronos.org/registry/OpenGL/extensions/KHR/KHR_blend_equation_advanced.txt) for GLES and avoid the extra render passes down the road.

**Review by bdero — APPROVED** [review]
Only minor nits and questions -- everything is looking great. Extremely excited to land this!

File: impeller/playground/backend/gles/playground_impl_gles.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (bdero) [reply impeller/playground/backend/gles/playground_impl_gles.cc:44]: Ah nvm, just noticed this was [already asked](https://github.com/flutter/engine/pull/33084#discussion_r864337490).

File: impeller/renderer/backend/gles/formats_gles.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (bdero) [reply impeller/renderer/backend/gles/formats_gles.h:145]: This was also [already pointed out](https://github.com/flutter/engine/pull/33084#discussion_r864386927)

File: impeller/renderer/formats.h (hunk 8 above)
Comment (chinmaygarde) [reply impeller/renderer/formats.h:90]: Thats a good point. From reading the spec, I believe `KHR_blend_equation_advanced_coherent` will be necessary right? In any case, I left these out because I could not find anything similar for Metal or Vulkan (and hadn't considered other API's like DX12). If we are going to need to work with these extensions anyway, I'd rather leave them out now and devise a separate strategy for extensions. I really don't want to be super prescriptive about not using extensions though. If they are available and avoid render passes, then I'm all for it. Could you please file a bug for this?

File: impeller/renderer/formats.h (hunk 7 above)
Comment (chinmaygarde) [reply impeller/renderer/formats.h:14]: Roger. I think clangd gets confused when the same file can be reached via multiple `-I` directives. I manually correct these but forgot to address this one.

File: impeller/renderer/backend/gles/device_buffer_gles.cc [hunk 9]
```diff
@@ -4,8 +4,103 @@
...
+                                   StorageMode mode)
+    : DeviceBuffer(size, mode),
+      reactor_(std::move(reactor)),
+      handle_(reactor_ ? reactor_->CreateHandle(HandleType::kBuffer)
+                       : GLESHandle::DeadHandle()) {}
+
+// |DeviceBuffer|
+DeviceBufferGLES::~DeviceBufferGLES() {
+  if (!reactor_) {
+    reactor_->CollectHandle(handle_);
```
[the remark this answers is not in the document]
Comment (chinmaygarde) [reply impeller/renderer/backend/gles/device_buffer_gles.cc:25]: Agreed. Fixed.

File: impeller/playground/backend/gles/playground_impl_gles.cc (hunk 1 above)
Comment (chinmaygarde) [reply impeller/playground/backend/gles/playground_impl_gles.cc:44]: Done.

File: impeller/renderer/backend/gles/formats_gles.h (hunk 2 above)
Comment (chinmaygarde) [reply impeller/renderer/backend/gles/formats_gles.h:145]: Done.

File: impeller/playground/imgui/imgui_impl_impeller.cc [hunk 10]
```diff
@@ -217,9 +215,13 @@ void ImGui_ImplImpeller_RenderDrawData(ImDrawData* draw_data,
...
-            std::max(0ll, clip_min.x), std::max(0ll, clip_min.y),
-            std::min(render_pass.GetRenderTargetSize().width, clip_max.x),
-            std::min(render_pass.GetRenderTargetSize().height, clip_max.y));
+            std::max<impeller::IRect::Type>(0ll, clip_min.x),  //
+            std::max<impeller::IRect::Type>(0ll, clip_min.y),  //
+            std::min<impeller::IRect::Type>(
+                render_pass.GetRenderTargetSize().width, clip_max.x),  //
+            std::min<impeller::IRect::Type>(
+                render_pass.GetRenderTargetSize().height, clip_max.y)  //
+        );
```
[the remark this answers is not in the document]
Comment (chinmaygarde) [reply impeller/playground/imgui/imgui_impl_impeller.cc:224]: I'd like to apply this in a separate patch if its alright.

File: impeller/renderer/backend/gles/vertex_descriptor_gles.h (hunk 3 above)
Comment (chinmaygarde) [reply impeller/renderer/backend/gles/vertex_descriptor_gles.h:18]: Renamed to BufferBindingsGLES

File: impeller/renderer/command.h (hunk 4 above)
Comment (chinmaygarde) [reply impeller/renderer/command.h:119]: Done.

File: impeller/renderer/formats.h (hunk 5 above)
Comment (chinmaygarde) [reply impeller/renderer/formats.h:190]: Done.

File: impeller/renderer/platform.h (hunk 6 above)
Comment (chinmaygarde) [reply impeller/renderer/platform.h:20]: Turns out this is necessary for GLES.

File: impeller/renderer/formats.h (hunk 8 above)
Comment (bdero) [reply impeller/renderer/formats.h:90]: SGTM, this is mentioned in https://github.com/flutter/flutter/issues/100086 but probably belongs in a separate issue.
